### PR TITLE
feat: admin scoring via CSV upload

### PR DESCRIPTION
**Pull request by zaki-amin**
In addition to using the form, admins can now upload a CSV for their scoring

**UI**
- Dropdown now includes an option for `ADMIN_SCORING`
  - Screenshot 2024-10-09 at 13 43 55

**Server**
- Admin scoring via CSV updates relevant tables and records the user/update time
- CSV uploads with user-friendly column names
- Refactor of schema validation

### CODE REVIEW

File: lib/schema.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  applicant: z.object({
+    cid: schemaCid,
+    ucasNumber: z.string().length(10, { message: 'UCAS number must be exactly 10 digits' }),
+    gender: z.nativeEnum(Gender),
+    firstName: z.string(),
+    surname: z.string(),
+    preferredName: z.string().nullable(),
+    dateOfBirth: z
+      .string()
+      .transform((value) => formatISO(parseDate(value, 'dd/MM/yyyy', new Date()))),
```
Comment (procaconsul) [inline lib/schema.ts:82]: Ok with the sttring format. If this one and/or the one on line 97 recur across different files, could be worth extracting them as common constants

**Review by procaconsul — CHANGES_REQUESTED** [review]

File: lib/csv/upload.ts [hunk 2]
```diff
@@ -2,8 +2,14 @@
...
 import prisma from '@/db'
 import { preprocessCsvData } from '@/lib/csv/preprocessing'
-import { parseWithSchema, schemaApplication, schemaTMUAScores, schemaUser } from '@/lib/csv/schema'
 import { allocateApplications } from '@/lib/reviewerAllocation'
+import {
+  parseWithSchema,
+  schemaCsvAdminScoring,
+  schemaCsvApplication,
+  schemaCsvTmuaScores,
+  schemaCsvUserRoles
```
[the remark this answers is not in the document]
Comment (zaki-amin) [reply lib/csv/upload.ts]: Postfixed with `schema` in f72c699c504c37f14d8ccee9ba6235de74024b13

File: lib/schema.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    .gte(from, { message: `${fieldName} must be ≥ ${from}` })
+    .lte(to, { message: `${fieldName} must be ≤ ${to}` })
+
+  return z.preprocess(
+    (val) => (val === '' ? null : Number(val)),
+    isNullable ? schema.nullable() : schema
+  )
+}
+
+const schemaCid = z.string().length(8, { message: 'CID must be exactly 8 characters' })
```
[the remark this answers is not in the document]
Comment (zaki-amin) [reply lib/schema.ts]: Field is more appropriate, resolved in f72c699c504c37f14d8ccee9ba6235de74024b13

File: lib/schema.ts (hunk 1 above)
Comment (zaki-amin) [reply lib/schema.ts:82]: These are the only two date formats and they're different so I'm inclined to leave these as they are

**Pull request merged**